### DISCUSSION TITLE
Allow for calling mutating interface methods on generic struct parameters

### DISCUSSION THREAD

**Discussion by Saalvage** [dotnet/csharplang#7402] [Language Ideas] [answered]
### The Problem
Assume we have the following setup:
```cs
class MyClass<T> where T : struct {
    T MyT;

    void PotentiallyModify(float newValue) {
        if (MyT is IModifyable m) {
            m.Modify(newValue);
        }
    }
}
```
This does not work as expected!

Since `T` is a struct, `is` results in `m` being a boxed copy of `MyT`, so any mutations being done are discarded!

### The Workaround
Manually maintaining a boxed instance of the struct allows for working around this:
```cs
class MyClass<T> where T : struct {
    private readonly object _myTBoxed = default(T);
    ref T MyT => ref Unsafe.Unbox<T>(_myTBoxed);

    void PotentiallyModify(float newValue) {
        if (_myTBoxed is IModifyable m) {
            m.Modify(newValue);
        }
    }
}
```
Works, but involves `Unsafe` and an extra field and is honestly pretty ugly..

### The Proposal
I propose something along the lines of the following to address this issue in a direct and easy manner:
```cs
if (MyT is ref IModifyable m) {
    m.Modify(newValue);
}
```
It would likely work by directly translating all calls to interface methods on `m` to the particular implementation in the struct, since the struct should be known by the compiler, meaning the entire `if` could be a compile-time check!

-----------------------------------
I recognize that this is likely a quite niche use case, but I think it would be a big step towards making structs and interfaces *interface* with one another much more nicely, as well as making generics slightly more powerful.

**Comment by Joe4evr** [comment, marked as answer]
Sounds like you want https://github.com/dotnet/csharplang/discussions/6308
That proposes the following:
```cs
if (T is IModifyable)
{
    // ANY instance of T in scope here can call IModifyable methods
}
```

**Reply by Windows10CE** [reply]
To show how this applies, it would allow you to do this in your example scenario:
```cs
class MyClass<T> where T : struct {
    T MyT;

    void PotentiallyModify(float newValue) {
        if (T is IModifyable) {
            MyT.Modify(newValue);
        }
    }
}
```
